### PR TITLE
Add requireCollectionFields flag

### PR DESCRIPTION
**Pull request by plokhotnyuk**

### CODE REVIEW

File: jsoniter-scala-macros/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/macros/JsonCodecMaker.scala [hunk 1]
```diff
@@ -931,8 +938,12 @@ object JsonCodecMaker {
           if (discriminator.isEmpty) names
           else names :+ n
         })
-        val required = classInfo.fields.collect {
-          case f if !f.symbol.isParamWithDefault && !isContainer(f.resolvedTpe) => f.mappedName
+        val required: Seq[String] = classInfo.fields.flatMap { f =>
```
Comment (jvican) [inline jsoniter-scala-macros/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/macros/JsonCodecMaker.scala]: Are you doing a deep check here? I was wondering would this logic behave in types such as `List[Option[Int]]` or `List[Set[Int]]`, would it force those fields to be present for the nested collections too?

File: jsoniter-scala-macros/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/macros/JsonCodecMaker.scala [hunk 2]
```diff
@@ -151,6 +156,7 @@ class CodecMakerConfig private (
       transientDefault = transientDefault,
       transientEmpty = transientEmpty,
       transientNone = transientNone,
+      requireCollectionFields = requireCollectionFields,
```
Comment (jvican) [inline jsoniter-scala-macros/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/macros/JsonCodecMaker.scala]: I'm not sure if we should consider `Option` included as a collection. It'd be weird that at least something marked as `Option[Int]` would be require if `requireCollectionFields = true`.

File: jsoniter-scala-macros/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/macros/JsonCodecMaker.scala (hunk 1 above)
Comment (plokhotnyuk) [reply jsoniter-scala-macros/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/macros/JsonCodecMaker.scala]: No deep checks - type parameters for arrays and iterables are not analyzed, so logic for working with the flag is the same for any kind of them, but logic of the flag work for all such kind of fields for nested data structures.

I've added more tests for codecs derived with `requireCollectionFields = true` to answer you questions. Please check if something still not clear about the flag.

File: jsoniter-scala-macros/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/macros/JsonCodecMaker.scala (hunk 2 above)
Comment (plokhotnyuk) [reply jsoniter-scala-macros/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/macros/JsonCodecMaker.scala]: This flag doesn't work for for fields which are optional or have defined default values.

**Comment by jvican** [comment]
What's the status of this change? It'd be lovely if I had it in a release I can use sometime in the next few days, where I plan to make the release.

**Comment by plokhotnyuk** [comment]
@jvican Hi, Jorge!

Thanks for your patience and support! 

I'm still in doubt to release this flag for the derivation macro in proposed implementation... 
Currently, raising of the `requireCollectionFields` flag requires turning off `transientEmpty` to make serialized messages be parsed back without throwing an error.

The main point that it can lead to misuse by not so experienced users and after releasing it will be very hard to rule derivation configuration to a better state.

So any advice about how to make it better upfront or easier to fix in the future will be greatly appreciated.

And, as W/A you can generate codecs with the snapshot version from this branch with turned on the `-Xmacro-settings:print-codecs` option for Scala compiler. Printed sources can be copy-pasted to code in place of the `make` macro calls after optional removing of `;` characters.

**Comment by jvican** [comment]
> So any advice about how to make it better upfront or easier to fix in the future will be greatly appreciated.
  
I suggest the following way of interacting with this option:
  
- Enabling `requireCollectionFields` turns off `transientEmpty` by default.
- The macro expander checks that when `requireCollectionFields` is enabled, `transientEmpty` is disabled, otherwise it throws an error.
  
Is this a satisfactory way of exposing the flag to users? I believe this behavior would make sense.

**Comment by plokhotnyuk** [comment]
@jvican thanks, I've picked 2nd option and will cut a release tomorrow morning.

**Comment by plokhotnyuk** [comment]
@jvican v1.1.0 is already available on the Maven Central:
http://repo1.maven.org/maven2/com/github/plokhotnyuk/jsoniter-scala/jsoniter-scala-macros_2.12/1.1.0/

**Comment by jvican** [comment]
Hurray, thanks for all the help and dedication @plokhotnyuk :+1:

**Comment by plokhotnyuk** [comment]
@jvican You are welcome, Jorge!

Please, take into account that `jsoniter-scala-core` has methods for reading/writing immediately from `java.io.InputStream`/`java.io.OutputStream` with internal buffering (no wrapping by the buffered streams is required) or from/to memory-mapped files through the `java.nio.ByteBuffer`. In both cases you don't need to keep a whole file in memory.

And, keep sending back your feedback, please!

**Pull request merged**